blaze/compute/sparksql.py

```diff
@@ -1,135 +1,88 @@
-        a = make_query(a)
-    if not isinstance(b, SparkSQLQuery):
-        b = make_query(b)
-    return compute_up(e, a, b, **kwargs)
 
+def make_sqlalchemy_table(expr):
+    name = expr._name
+    columns = dshape_to_alchemy(expr.dshape)
+    return sa.Table(name, sa.MetaData(), *columns)
 
-@dispatch((UnaryOp, Expr), SparkSQLQuery)
-def compute_up(expr, q, **kwargs):
-    scope = kwargs.pop('scope', dict())
-    scope = dict((t, q.mapping.get(data, data)) for t, data in scope.items())
 
-    q2 = compute_up(expr, q.query, scope=scope, **kwargs)
-    return SparkSQLQuery(q.context, q2, q.mapping)
+@curry
+def istable(db, t):
+    return (isinstance(t, Field) and isrecord(t.dshape.measure) and
+            t._child.isidentical(db))
 
 
-@dispatch((BinOp, Join, Expr), SparkSQLQuery, SparkSQLQuery)
-def compute_up(expr, a, b, **kwargs):
-    assert a.context == b.context
+@dispatch(Expr, SQLContext)
+def compute_down(expr, data):
+    """ Compile a blaze expression to a sparksql expression"""
+    leaves = expr._leaves()
 
-    mapping = toolz.merge(a.mapping, b.mapping)
+    # make sure we only have a single leaf node
+    if len(leaves) != 1:
+        raise ValueError('Must compile from exactly one root database')
 
-    scope = kwargs.pop('scope', dict())
-    scope = dict((t, mapping.get(data, data)) for t, data in scope.items())
+    leaf, = leaves
 
-    c = compute_up(expr, a.query, b.query, scope=scope, **kwargs)
-    return SparkSQLQuery(a.context, c, mapping)
+    # field expressions on the database are Field instances with a record
+    # measure whose immediate child is the database leaf
+    tables = pipe(expr._subterms(), filter(istable(leaf)), list)
 
+    # raise if we don't have tables in our database
+    if not tables:
+        raise ValueError('Expressions not referencing a table cannot be '
+                         'compiled')
 
-from .sql import select
-def sql_string(query):
-    return pipe(query, select, literalquery, str)
+    # make new symbols for each table
+    new_leaves = [symbol(t._name, t.dshape) for t in tables]
 
-@dispatch(Expr, SparkSQLQuery)
-def post_compute(expr, query, scope=None):
-    result = query.context.sql(sql_string(query.query))
-    if iscollection(expr.dshape) and isscalar(expr.dshape.measure):
-        result = result.map(lambda x: x[0])
-    return result
+    # sub them in the expression
+    expr = expr._subs(dict(zip(tables, new_leaves)))
 
+    # compute using sqlalchemy
+    scope = dict(zip(new_leaves, map(make_sqlalchemy_table, tables)))
+    query = compute(expr, scope)
 
-@dispatch(Head, SparkSQLQuery)
-def post_compute(expr, query, scope=None):
-    result = query.context.sql(sql_string(query.query))
-    if iscollection(expr.dshape) and isscalar(expr.dshape.measure):
-        result = result.map(lambda x: x[0])
-    return result.collect()
+    # interpolate params
+    compiled = literalquery(query, dialect=HiveDialect())
+    return data.sql(str(compiled))
```

Comment:
I've probably said this already, but this `compute_down` implementation is really slick.